### DISCUSSION TITLE
Why does the open method triggered when opening a WebSocket connection in the code example retrieve a custom structure, PerSocketData, which isn't passed by the client when the connection is opened?

### DISCUSSION THREAD

**Discussion by LetAmericaGreatAgain** [uNetworking/uWebSockets#1824] [Q&A]
**Why does the open method triggered when opening a WebSocket connection in the code example retrieve a custom structure, PerSocketData, which isn't passed by the client when the connection is opened? Since it is possible to directly access this structure through the ws->getUserData() method on the server side without any user initialization, why isn't this structure embedded into the code library, instead of being user-defined? This really confuses me.**

```c++
.open = [](uWS::WebSocket<false,true, PerSocketData> *ws) {
                /* Open event here, you may access ws->getUserData() which points to a PerSocketData struct */

                PerSocketData *perSocketData = (PerSocketData *) ws->getUserData();

                for (int i = 0; i < 32; i++) {
                    std::string topic = std::to_string((uintptr_t)ws) + "-" + std::to_string(i);
                    perSocketData->topics.push_back(topic);
                    ws->subscribe(topic);
                }
            }
```

**Comment by uNetworkingAB** [comment]
PerSocketData has no relation to the client making the connection. PerSocketData is data you need on the server side, per socket. And it's defined by you, as you need it. You init it as you need it, inside .open and use it as you wish

**Reply by LetAmericaGreatAgain** [reply]
Thank you, I understand now. This is an extension of the socket, and the codebase likely creates an instance of the custom socket extension type based on the template type parameter during WebSocket initialization. When the connection is opened, the open event can directly access it.
